evm/src/main/java/org/hyperledger/besu/evm/EVM.java

```diff
@@ -70,59 +70,43 @@ public GasCalculator getGasCalculator() {
   }
 
   public void runToHalt(final MessageFrame frame, final OperationTracer operationTracer) {
+    byte[] code = frame.getCode().getBytes().toArrayUnsafe();
     while (frame.getState() == MessageFrame.State.CODE_EXECUTING) {
-      executeNextOperation(frame, operationTracer);
-    }
-  }
-
-  private void executeNextOperation(
-      final MessageFrame frame, final OperationTracer operationTracer) {
-    frame.setCurrentOperation(operationAtOffset(frame.getCode(), frame.getPC()));
-    operationTracer.traceExecution(
-        frame,
-        () -> {
-          OperationResult result;
-          final Operation operation = frame.getCurrentOperation();
-          try {
-            result = operation.execute(frame, this);
-          } catch (final OverflowException oe) {
-            result = OVERFLOW_RESPONSE;
-          } catch (final UnderflowException ue) {
-            result = UNDERFLOW_RESPONSE;
-          }
-          logState(frame, result.getGasCost().orElse(0L));
-          final Optional<ExceptionalHaltReason> haltReason = result.getHaltReason();
-          if (haltReason.isPresent()) {
-            LOG.trace("MessageFrame evaluation halted because of {}", haltReason.get());
-            frame.setExceptionalHaltReason(haltReason);
-            frame.setState(State.EXCEPTIONAL_HALT);
-          } else if (result.getGasCost().isPresent()) {
-            frame.decrementRemainingGas(result.getGasCost().getAsLong());
-          }
-          if (frame.getState() == State.CODE_EXECUTING) {
-            final int currentPC = frame.getPC();
-            final int opSize = result.getPcIncrement();
-            frame.setPC(currentPC + opSize);
-          }
-
-          return result;
-        });
-  }
+        byte opcode;
```

Comment:
Might be worth a comment here to say it's been deliberately inlined for performance reasons. Otherwise someone's likely to extract a method again to make the code cleaner.